zip-0315.rst

```diff
@@ -1,9 +1,629 @@
 ::
 
   ZIP: 315
-  Title: Best Practices for Wallet Handling of Multiple Pools
+  Title: Best Practices for Wallet Implementations
   Owners: Daira Emma Hopwood <daira@electriccoin.co>
           Jack Grigg <jack@electriccoin.co>
-  Status: Reserved
+          Kris Nuttycombe <kris@electriccoin.co>
+  Credits: Francisco Gindre
+  Status: Proposed
   Category: Wallet
   Discussions-To: <https://github.com/zcash/zips/issues/447>
+  Pull-Request: <https://github.com/zcash/zips/issues/>
+
+
+Terminology
+===========
+
+The key words "MUST", "SHOULD", "SHOULD NOT", and "MAY" in this document are to be
+interpreted as described in RFC 2119. [#RFC2119]_
+
+The terms below are to be interpreted as follows:
+
+Auto-shielding
+    The process of automatically transferring transparent funds to the most recent
+    shielded pool on receipt.
+
+Opportunistic shielding
+    The process of transferring previously received transparent funds to the most
+    recent shielded pool as part of a user-initiated transaction.
+
+Auto-migration
+    The process of automatically transferring shielded funds from older pools to the
+    most preferred (usually the most recent) pool on receipt.
+
+Opportunistic migration
+    The process of transferring previously received shielded funds from older pools
+    to the most preferred (usually the most recent) pool as part of a user-initiated
+    transaction.
```

Comment:
Opportunistic migration has a similar problem to opportunistic shielding: it links the payment to the amount being transferred between pools, from the point of view of a recipient or IVK holder.

If, say, all of your funds are currently in Sapling, and the next transaction after you enable opportunistic migration (or after your wallet makes it the default) happens to be to an untrusted party, then a naive implementation of opportunistic shielding will reveal your total wealth in ZEC to that party.

I'm open to opinions on whether we should just recommend against opportunistic migration in the same way as opportunistic shielding, or just document the problem and perhaps recommend a limit on the amount migrated in each tx. I think we will need to do so in such a way that we don't completely rule out opportunistic balancing of funds between pools.

Reply:
I agree with @daira has stated. I think that the "opportunity" has to be consented by the user on a _per transaction_ basis. 

Not only there's a big chance of correlating balances to people when doing a transaction with a third party, but also I can recalls problems with a past **coin voting** and ZecWallet opportunistic shielding. There might be times were someone sets a funds for a purpose somewhere and forgets about it in another context. It's not that the person does not care about those funds being there anymore, it's just how the human brain works.  

If the wallet doesn't state explicitly what's the "opportunity" then the user might not recall having delegated that decision to the wallet and unintended things can happen.